lib/aio/github.py

```diff
@@ -0,0 +1,179 @@
+                    logger.debug('  cache hit %r -- returning cached value', resource)
+                    return cache_entry.value
+
+                else:
+                    value = await response.json()
+                    logger.debug('  cache miss %r -- caching and returning %r', resource, conditions)
+                    self.cache.add(resource, CacheEntry(conditions, value))
+                    return value  # type: ignore[no-any-return]
+
+        for attempt in range(4):
+            try:
+                return await get_once()
+            except aiohttp.ClientResponseError as exc:
+                if exc.status < 500:
+                    raise
+            except aiohttp.ClientError:
+                pass
+
+            # 1 → 2 → 4 → 8s delay
+            await asyncio.sleep(2 ** attempt)
+
+        # ...last attempt.
+        return await get_once()
+
+    async def check_pr_changed(self, repo: str, pull_nr: int, expected_sha: str) -> str | None:
+        try:
+            pull = typechecked(await self.get(f'repos/{repo}/pulls/{pull_nr}'), dict)
+            if get_str(pull, 'state') != 'open':
+                return f'{repo}#{pull_nr} is closed'
+            if get_str(get_dict(pull, 'head'), 'sha') != expected_sha:
+                return f'{repo}#{pull_nr} changed'
+        except JsonError as exc:
+            return f'Unexpected error when parsing pull request: {exc}'
+        except aiohttp.ClientError as exc:
+            # might be transient, so don't kill the job on account of this...
+            logger.warning('Error when polling for %s#%s: %r', repo, pull_nr, exc)
+            return None
+        else:
+            return None
+
+    async def open_issue(self, repo: str, issue: JsonObject) -> None:
+        await self.post(f'repos/{repo}/issues', issue)
+
+    def get_status(self, repo: str, sha: str, context: str, location: URL) -> Status:
+        return GitHubStatus(self, repo, sha, context, location)
+
+    async def resolve_subject(
+        self, repo: str, sha: str | None, pull_nr: int | None, branch: str | None, target: str | None
+    ) -> Subject:
+        clone_url = self.clone / repo
+
+        if pull_nr is not None:
+            pull = typechecked(await self.get(f'repos/{repo}/pulls/{pull_nr}'), dict)
+
+            # if our sha doesn't match the PR, we will detect it soon
+            target = get_str(get_dict(pull, 'base'), 'ref')
+            if sha is not None:
+                return Subject(clone_url, sha, target)
+            else:
+                return Subject(clone_url, get_str(get_dict(pull, 'head'), 'sha'), target)
+
+        else:
+            sha = await resolve_branch(self.session, clone_url, branch)
+            return Subject(clone_url, sha, target)
+
+
+class GitHubStatus(Status):
+    def __init__(self, api: GitHub, repo: str, revision: str, context: str, link: URL) -> None:
+        logger.debug('GitHub repo %s context %s link %s', repo, context, link)
```

Comment:
why not `revision`? That's no less important than repo and context.

Reply:
OK.